File: go/cars-assemble/cars_assemble.go
package cars

// CalculateWorkingCarsPerHour calculates how many working cars are
// produced by the assembly line every hour.
func CalculateWorkingCarsPerHour(productionRate int, successRate float64) float64 {
	return float64(productionRate) * successRate / 100.0
}

// CalculateWorkingCarsPerMinute calculates how many working cars are
// produced by the assembly line every minute.
func CalculateWorkingCarsPerMinute(productionRate int, successRate float64) int {
	return int(CalculateWorkingCarsPerHour(productionRate, successRate)) / 60
}

// costPerUnitCar: production cost of a single car
const costPerUnitCar = 10000
// costPerTenCars: production cost of ten cars
const costPerTenCars = 95000

// CalculateCost works out the cost of producing the given number of cars.
func CalculateCost(carsCount int) uint {
    totalCost := 0
	totalCost += costPerTenCars * (carsCount / 10)
    totalCost += costPerUnitCar * (carsCount % 10)
	return uint(totalCost)
}

File: lua/cars-assemble/cars-assemble.lua
local GROUP_SIZE <const> = 10
local COST <const> = {['unit'] = 10000, ['group'] = 95000}

local cars = {}

-- returns the amount of working cars produced by the assembly line every hour
function cars.calculate_working_cars_per_hour(production_rate, success_rate)
  return production_rate * success_rate / 100
end

-- returns the amount of working cars produced by the assembly line every minute
function cars.calculate_working_cars_per_minute(production_rate, success_rate)
  return cars.calculate_working_cars_per_hour(production_rate, success_rate) // 60
end

-- returns the cost of producing the given number of cars
function cars.calculate_cost(cars_count)
  local groups = cars_count // GROUP_SIZE
  local units = cars_count % GROUP_SIZE
  return groups * COST.group + units * COST.unit
end

return cars
